codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
         return notNull
     }
 
+    private fun mapFunctionLogic(
+        expression: JmespathExpression,
+        arg: VisitedExpression,
+        resultName: String,
+        operation: String,
+        stringOrNumberCheck: Boolean = true,
+    ): String {
+        val argName = arg.identifier
+        val result = bestTempVarName(resultName)
+        writer.withBlock("val $result = $argName?.$operation{", "}") {
+            val expressionValue = addTempVar("expression", subfieldCodegen((expression as ExpressionTypeExpression).expression as FieldExpression, "it"))
+            if (stringOrNumberCheck) {
+                write("if ($expressionValue as Any !is Number && $expressionValue as Any !is String) throw Exception(\"Result of applying expression should be string or number\")")
```

Comment:
**correctness**:

This check is wrong:

```kotlin
            val sorted = structs?.sortedBy{
                val expression = it?.integer
                if (expression as Any !is Number && expression as Any !is String) throw Exception("Result of applying expression should be string or number")
                expression
            }
```

`expression` is of type `Any?` so `as Any` is incorrect. Expression is likely never going to be null anyway though so I guess in practice it will be `Any` at runtime. If we can we should generate the code as `it.integer` not `it?` here since `.sortedBy` will only run if `structs` isn't null in the first place (due to `?`). 

Backing up for a second, why do we even need to generate the `stringOrNumber` check in the first place? I see in the [JMESPath spec](https://jmespath.org/specification.html#sort-by):

> If the result of evaluating the expr against the current array element results in type other than a number or a string, an invalid-type error will occur.

But why do we need to worry about this? Seems like this would be an invalid model most likely.

Reply:
> But why do we need to worry about this? Seems like this would be an invalid model most likely.

The expression resulting in anything other than a string or number is not being checked for at the model level. I think it makes sense to have that check. Someone could add an expression that results in something else and it would fail without an exception explaining why

Reply:
Also Smithy is doing some validation on the JMESPath expressions in the model: https://github.com/smithy-lang/smithy/blob/main/smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/TypeChecker.java

I don't think we need to add runtime validations here, this is the kind of stuff that belongs as an error in the model such that it never can even get into the hands of a customer.